### ISSUE TITLE
WARNING: possible circular locking dependency detected

### ISSUE THREAD

**Issue by marcin-github** [koverstreet/bcachefs#156]
Hi!

```

[ 2135.911421] ======================================================
[ 2135.911693] WARNING: possible circular locking dependency detected
[ 2135.911960] 5.9.0-00327-gac76b6926cd6 #46 Tainted: G           OE   T
[ 2135.912239] ------------------------------------------------------
[ 2135.912502] collectd/3475 is trying to acquire lock:
[ 2135.912738] ffff9b76c7ca9ae0 (&mm->mmap_lock#2){++++}-{3:3}, at: __access_remote_vm+0x4d/0x200
[ 2135.913175]
               but task is already holding lock:
[ 2135.913508] ffffffff9a06e320 (fs_reclaim){+.+.}-{0:0}, at: fs_reclaim_acquire.part.0+0x5/0x30
[ 2135.913925]
               which lock already depends on the new lock.

[ 2135.914393]
               the existing dependency chain (in reverse order) is:
[ 2135.914753]
               -> #1 (fs_reclaim){+.+.}-{0:0}:
[ 2135.915087]        fs_reclaim_acquire.part.0+0x25/0x30
[ 2135.915327]        __alloc_pages_nodemask+0x256/0x420
[ 2135.915574]        __pmd_alloc+0x2c/0x180
[ 2135.915817]        move_page_tables+0x885/0x940                                                                                                                                                            [ 2135.916084]        shift_arg_pages+0xad/0x190
[ 2135.916305]        setup_arg_pages+0x22a/0x260
[ 2135.916546]        load_elf_binary+0x1e0/0x1560
[ 2135.916783]        bprm_execve+0x368/0x880
[ 2135.917008]        kernel_execve+0x146/0x1d0
[ 2135.917275]        call_usermodehelper_exec_async+0xd0/0x140
[ 2135.917550]        ret_from_fork+0x22/0x30
[ 2135.917762]
               -> #0 (&mm->mmap_lock#2){++++}-{3:3}:
[ 2135.918129]        __lock_acquire+0x1219/0x23c0
[ 2135.918365]        lock_acquire+0xa6/0x370
[ 2135.918601]        down_read_killable+0x32/0x160
[ 2135.918831]        __access_remote_vm+0x4d/0x200
[ 2135.919077]        proc_pid_cmdline_read+0x21e/0x400
[ 2135.919316]        vfs_read+0x9f/0x190
[ 2135.919520]        ksys_read+0x65/0xe0
[ 2135.919728]        do_syscall_64+0x33/0x80
[ 2135.919945]        entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 2135.920201]
               other info that might help us debug this:

[ 2135.920699]  Possible unsafe locking scenario:

[ 2135.921048]        CPU0                    CPU1
[ 2135.922013]        ----                    ----
[ 2135.922280]   lock(fs_reclaim);
[ 2135.922560]                                lock(&mm->mmap_lock#2);
[ 2135.922822]                                lock(fs_reclaim);
[ 2135.923043]   lock(&mm->mmap_lock#2);
[ 2135.923221]
                *** DEADLOCK ***

[ 2135.923640] 2 locks held by collectd/3475:
[ 2135.923871]  #0: ffff9b77223ca350 (&f->f_pos_lock){+.+.}-{3:3}, at: __fdget_pos+0x47/0x50
[ 2135.924375]  #1: ffffffff9a06e320 (fs_reclaim){+.+.}-{0:0}, at: fs_reclaim_acquire.part.0+0x5/0x30
[ 2135.924755]
               stack backtrace:
[ 2135.925052] CPU: 3 PID: 3475 Comm: collectd Tainted: G           OE   T 5.9.0-00327-gac76b6926cd6 #46
[ 2135.925449] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS rel-1.13.0-48-gd9c812dda519-prebuilt.qemu.org 04/01/2014
[ 2135.925904] Call Trace:
[ 2135.926142]  dump_stack+0x77/0xa0
[ 2135.926346]  check_noncircular+0x162/0x180
[ 2135.926578]  __lock_acquire+0x1219/0x23c0
[ 2135.926799]  lock_acquire+0xa6/0x370
[ 2135.927015]  ? __access_remote_vm+0x4d/0x200
[ 2135.927226]  ? __schedule+0x4e0/0xa90
[ 2135.927420]  down_read_killable+0x32/0x160
[ 2135.927643]  ? __access_remote_vm+0x4d/0x200
[ 2135.927870]  __access_remote_vm+0x4d/0x200
[ 2135.928098]  ? __alloc_pages_nodemask+0x3c7/0x420
[ 2135.928343]  proc_pid_cmdline_read+0x21e/0x400
[ 2135.928579]  vfs_read+0x9f/0x190
[ 2135.928790]  ksys_read+0x65/0xe0
[ 2135.928997]  do_syscall_64+0x33/0x80
[ 2135.929210]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 2135.929461] RIP: 0033:0x7f1e62b6212c
[ 2135.929677] Code: ec 28 48 89 54 24 18 48 89 74 24 10 89 7c 24 08 e8 89 fc ff ff 41 89 c0 48 8b 54 24 18 48 8b 74 24 10 8b 7c 24 08 31 c0 0f 05 <48> 3d 00 f0 ff ff 77 30 44 89 c7 48 89 44 24 08 e8 bf fc ff ff 48
[ 2135.930329] RSP: 002b:00007f1e4f7fb310 EFLAGS: 00000246 ORIG_RAX: 0000000000000000
[ 2135.930682] RAX: ffffffffffffffda RBX: 0000000000000000 RCX: 00007f1e62b6212c
[ 2135.931048] RDX: 0000000000001000 RSI: 00007f1e4f7fbe20 RDI: 0000000000000013
[ 2135.931409] RBP: 00007f1e4f7fbe20 R08: 0000000000000000 R09: 00007f1e4f7fb1d0
[ 2135.931752] R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000001000
[ 2135.932117] R13: 0000000000000013 R14: 00007f1e4f7fe698 R15: 00007f1e4f7fce20
```

**Comment by marcin-github** [comment]
```
# scripts/faddr2line vmlinux __access_remote_vm+0x4d/0x200 fs_reclaim_acquire.part.0+0x5/0x30  __fdget_pos+0x47/0x50
__access_remote_vm+0x4d/0x200:
__access_remote_vm at mm/memory.c:4880

fs_reclaim_acquire.part.0+0x5/0x30:
fs_reclaim_acquire.part.0 at mm/page_alloc.c:4234

__fdget_pos+0x47/0x50:
__fdget_pos at fs/file.c:930
```

**Comment by marcin-github** [comment]
Another deadlock with something new (I don't know if it's useful)
```
[22329.510822] BUG: workqueue leaked lock or atomic: kworker/4:0/0x00000000/11530
                    last function: update_balloon_size_func [virtio_balloon]
[22329.510864] 1 lock held by kworker/4:0/11530:
[22329.510866]  #0: ffffffffa506e320 (fs_reclaim){+.+.}-{0:0}, at: fs_reclaim_acquire.part.0+0x5/0x30
[22329.510902] CPU: 4 PID: 11530 Comm: kworker/4:0 Tainted: G           OE   T 5.9.0-00327-gac76b6926cd6 #46
[22329.510904] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS rel-1.13.0-48-gd9c812dda519-prebuilt.qemu.org 04/01/2014
[22329.510908] Workqueue: events_freezable update_balloon_size_func [virtio_balloon]
[22329.510911] Call Trace:
[22329.510923]  dump_stack+0x77/0xa0
[22329.510932]  process_one_work.cold+0x41/0x46
[22329.510935]  worker_thread+0x50/0x390
[22329.510938]  ? process_one_work+0x560/0x560
[22329.510941]  kthread+0x112/0x150
[22329.510943]  ? kthread_park+0x90/0x90
[22329.510950]  ret_from_fork+0x22/0x30
[22329.511100] BUG: workqueue leaked lock or atomic: kworker/4:0/0x00000000/11530
                    last function: drain_local_pages_wq
[22329.511115] 1 lock held by kworker/4:0/11530:
[22329.511117]  #0: ffffffffa506e320 (fs_reclaim){+.+.}-{0:0}, at: fs_reclaim_acquire.part.0+0x5/0x30
[22329.511125] CPU: 4 PID: 11530 Comm: kworker/4:0 Tainted: G           OE   T 5.9.0-00327-gac76b6926cd6 #46
[22329.511127] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS rel-1.13.0-48-gd9c812dda519-prebuilt.qemu.org 04/01/2014
[22329.511129] Workqueue: mm_percpu_wq drain_local_pages_wq
[22329.511131] Call Trace:
[22329.511135]  dump_stack+0x77/0xa0
[22329.511138]  process_one_work.cold+0x41/0x46
[22329.511141]  worker_thread+0x1d8/0x390
[22329.511144]  ? process_one_work+0x560/0x560
[22329.511163]  kthread+0x112/0x150
[22329.511167]  ? kthread_park+0x90/0x90
[22329.511169]  ret_from_fork+0x22/0x30
[22329.511423] BUG: workqueue leaked lock or atomic: kworker/4:0/0x00000000/11530
                    last function: wq_barrier_func
[22329.511426] 1 lock held by kworker/4:0/11530:
[22329.511427]  #0: ffffffffa506e320 (fs_reclaim){+.+.}-{0:0}, at: fs_reclaim_acquire.part.0+0x5/0x30
[22329.511435] CPU: 4 PID: 11530 Comm: kworker/4:0 Tainted: G           OE   T 5.9.0-00327-gac76b6926cd6 #46
[22329.511437] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS rel-1.13.0-48-gd9c812dda519-prebuilt.qemu.org 04/01/2014
[22329.511439] Workqueue: mm_percpu_wq wq_barrier_func
[22329.511442] Call Trace:
[22329.511445]  dump_stack+0x77/0xa0
[22329.511449]  process_one_work.cold+0x41/0x46
[22329.511452]  worker_thread+0x1d8/0x390
[22329.511455]  ? process_one_work+0x560/0x560
[22329.511457]  kthread+0x112/0x150
[22329.511460]  ? kthread_park+0x90/0x90
[22329.511462]  ret_from_fork+0x22/0x30

[22329.511701] ======================================================
[22329.511967] WARNING: possible circular locking dependency detected
[22329.512229] 5.9.0-00327-gac76b6926cd6 #46 Tainted: G           OE   T
[22329.512503] ------------------------------------------------------
[22329.512775] kworker/4:0/11530 is trying to acquire lock:
[22329.513028] ffff9a2f7a059338 ((wq_completion)events_freezable){+.+.}-{0:0}, at: process_one_work+0x1af/0x560
[22329.513438]
               but task is already holding lock:
[22329.513782] ffffffffa506e320 (fs_reclaim){+.+.}-{0:0}, at: fs_reclaim_acquire.part.0+0x5/0x30
[22329.514172]
               which lock already depends on the new lock.

[22329.514644]
               the existing dependency chain (in reverse order) is:
[22329.515024]
               -> #1 (fs_reclaim){+.+.}-{0:0}:
[22329.515372]        lock_release+0xf2/0x260
[22329.515607]        process_one_work+0x25d/0x560
[22329.526115]        worker_thread+0x50/0x390
[22329.526340]        kthread+0x112/0x150
[22329.526550]        ret_from_fork+0x22/0x30
[22329.526765]
               -> #0 ((wq_completion)events_freezable){+.+.}-{0:0}:
[22329.527130]        __lock_acquire+0x1219/0x23c0
[22329.527360]        lock_acquire+0xa6/0x370
[22329.527554]        process_one_work+0x1e6/0x560
[22329.527778]        worker_thread+0x50/0x390
[22329.527985]        kthread+0x112/0x150
[22329.528174]        ret_from_fork+0x22/0x30
[22329.528363]
               other info that might help us debug this:

[22329.529953]  Possible unsafe locking scenario:

[22329.530291]        CPU0                    CPU1
[22329.530515]        ----                    ----
[22329.530737]   lock(fs_reclaim);
[22329.530932]                                lock((wq_completion)events_freezable);
[22329.531291]                                lock(fs_reclaim);
[22329.531540]   lock((wq_completion)events_freezable);
[22329.531776]
                *** DEADLOCK ***

[22329.532198] 1 lock held by kworker/4:0/11530:
[22329.532424]  #0: ffffffffa506e320 (fs_reclaim){+.+.}-{0:0}, at: fs_reclaim_acquire.part.0+0x5/0x30
[22329.532820]
               stack backtrace:
[22329.533139] CPU: 4 PID: 11530 Comm: kworker/4:0 Tainted: G           OE   T 5.9.0-00327-gac76b6926cd6 #46
[22329.533543] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS rel-1.13.0-48-gd9c812dda519-prebuilt.qemu.org 04/01/2014
[22329.533976] Workqueue: events_freezable update_balloon_size_func [virtio_balloon]
[22329.534315] Call Trace:
[22329.534506]  dump_stack+0x77/0xa0
[22329.534709]  check_noncircular+0x162/0x180
[22329.534932]  __lock_acquire+0x1219/0x23c0
[22329.535156]  lock_acquire+0xa6/0x370
[22329.535364]  ? process_one_work+0x1af/0x560
[22329.535583]  ? mark_held_locks+0x46/0x70
[22329.535801]  process_one_work+0x1e6/0x560
[22329.536029]  ? process_one_work+0x1af/0x560
[22329.536247]  worker_thread+0x50/0x390
[22329.536485]  ? process_one_work+0x560/0x560
[22329.536684]  kthread+0x112/0x150
[22329.536873]  ? kthread_park+0x90/0x90
[22329.545906]  ret_from_fork+0x22/0x30
[22338.535171] virtio_balloon virtio0: Out of puff! Can't get 1 pages
[22348.687981] virtio_balloon virtio0: Out of puff! Can't get 1 pages
```

**Comment by koverstreet** [comment]
Probably also fixed by 2f6fb2b0b7
